Implement Hash if members impl Hash

pitaj opened the pull request:
Not sure if `extra` should be included in this or not.

ProgVal commented:
What is the motivation for this?

`extra` must not be included, see the `PartialEq` implementation and https://doc.rust-lang.org/std/hash/trait.Hash.html#hash-and-eq

And could you add tests?

pitaj commented:
Motivation is that I have several types like the following:
```rust
#[derive(Debug, PartialEq, Eq, Clone, Hash)]
pub enum Token<'a> {
    // Template text passed through
    Text(Span<'a>),
    ...
}
```
Every variant holds a span. If `LocatedSpan` implements Hash, I can just `derive(Hash)`, otherwise one must `impl Hash` for each type which has a span.

> extra must not be included

Yeah I figured that. Fixed.

> And could you add tests?

I added one, mirroring the equality one. Let me know if you'd like more.

ProgVal commented:
alright, thanks!
